Mathlib/Data/Fin/Tuple/NatAntidiagonal.lean

```diff
@@ -0,0 +1,285 @@
+Authors: Eric Wieser
+
+! This file was ported from Lean 3 source module data.fin.tuple.nat_antidiagonal
+! leanprover-community/mathlib commit 98e83c3d541c77cdb7da20d79611a780ff8e7d90
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.BigOperators.Fin
+import Mathlib.Data.Finset.NatAntidiagonal
+import Mathlib.Data.Fin.VecNotation
+import Mathlib.Logic.Equiv.Fin
+
+/-!
+# Collections of tuples of naturals with the same sum
+
+This file generalizes `List.Nat.Antidiagonal n`, `Multiset.Nat.Antidiagonal n`, and
+`Finset.Nat.Antidiagonal n` from the pair of elements `x : ℕ × ℕ` such that `n = x.1 + x.2`, to
+the sequence of elements `x : Fin k → ℕ` such that `n = ∑ i, x i`.
+
+## Main definitions
+
+* `List.Nat.antidiagonalTuple`
+* `Multiset.Nat.antidiagonalTuple`
+* `Finset.Nat.antidiagonalTuple`
+
+## Main results
+
+* `antidiagonalTuple 2 n` is analogous to `antidiagonal n`:
+
+  * `List.Nat.antidiagonalTuple_two`
+  * `Multiset.Nat.antidiagonalTuple_two`
+  * `Finset.Nat.antidiagonalTuple_two`
+
+## Implementation notes
+
+While we could implement this by filtering `(Fintype.PiFinset $ λ _, range (n + 1))` or similar,
+this implementation would be much slower.
+
+In the future, we could consider generalizing `Finset.Nat.antidiagonalTuple` further to
+support finitely-supported functions, as is done with `cut` in
+`archive/100-theorems-list/45_partition.lean`.
+-/
+
+
+open BigOperators
+
+/-! ### Lists -/
+
+
+namespace List.Nat
+
+/-- `List.antidiagonalTuple k n` is a list of all `k`-tuples which sum to `n`.
+
+This list contains no duplicates (`List.Nat.nodup_antidiagonalTuple`), and is sorted
+lexicographically (`List.Nat.antidiagonalTuple_pairwise_pi_lex`), starting with `![0, ..., n]`
+and ending with `![n, ..., 0]`.
+
+```
+#eval antidiagonal_tuple 3 2
+-- [![0, 0, 2], ![0, 1, 1], ![0, 2, 0], ![1, 0, 1], ![1, 1, 0], ![2, 0, 0]]
+```
+-/
+def antidiagonalTuple : ∀ k, ℕ → List (Fin k → ℕ)
+  | 0, 0 => [![]]
+  | 0, _ + 1 => []
+  | k + 1, n =>
+    (List.Nat.antidiagonal n).bind fun ni =>
+      (antidiagonalTuple k ni.2).map fun x => Fin.cons ni.1 x
+#align list.nat.antidiagonal_tuple List.Nat.antidiagonalTuple
+
+@[simp]
+theorem antidiagonalTuple_zero_zero : antidiagonalTuple 0 0 = [![]] :=
+  rfl
+#align list.nat.antidiagonal_tuple_zero_zero List.Nat.antidiagonalTuple_zero_zero
+
+@[simp]
+theorem antidiagonalTuple_zero_succ (n : ℕ) : antidiagonalTuple 0 n.succ = [] :=
+  rfl
+#align list.nat.antidiagonal_tuple_zero_succ List.Nat.antidiagonalTuple_zero_succ
+
+theorem mem_antidiagonalTuple {n : ℕ} {k : ℕ} {x : Fin k → ℕ} :
+    x ∈ antidiagonalTuple k n ↔ (∑ i, x i) = n := by
+  induction x using Fin.consInduction generalizing n with
+  | h0 =>
+    cases n
+    · simp
+    · simp [eq_comm]
+  | h x₀ x ih =>
+    simp_rw [Fin.sum_cons]
+    rw [antidiagonalTuple]  -- porting note: simp_rw doesn't use the equation lemma properly
```

Comment:
This is worth raising on Zulip

Reply:
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/simp_rw.20not.20working.20as.20expected